dlt/extract/source.py

```diff
@@ -614,6 +614,15 @@ def root_key(self) -> bool:
         config = RelationalNormalizer.get_normalizer_config(self._schema).get("propagation")
         return config is not None and "root" in config and "_dlt_id" in config["root"] and config["root"]["_dlt_id"] == "_dlt_root_id"
 
+    @property
+    def exhausted(self) -> bool:
+        # check all selected pipes wether one of them is exhausted
+        for resource in self._resources.extracted.values():
+            for item in resource._pipe:
```

Comment:
you do not need to enumerate items. `Pipe` has `gen` property which is the main generator of the pipe. all others are transforms. so check only `gen`